src/rateLimiters/fixedWindow.ts

```diff
@@ -0,0 +1,129 @@
+        this.capacity = capacity;
+        this.windowSize = windowSize;
+        this.client = client;
+        if (windowSize <= 0 || capacity <= 0)
+            throw Error('FixedWindow windowSize and capacity must be positive');
+    }
+
+    /**
+     * @function processRequest - Fixed Window algorithm to allow or block
+     * based on the depth/complexity (in amount of tokens) of incoming requests.
+     *               Fixed Window
+     *     _________________________________
+     *    |         *full capacity          |
+     *    |                                 |  move to next time window
+     *    |    token adds up until full     |       ---------->
+     *____._________________________________.____
+     *    |<--        window size        -->|
+     *current timestamp              next timestamp
+     *
+     * First, checks if a window exists in the redis cache.
+     * If not, then `fixedWindowStart` is set as the current timestamp, and `currentTokens` is checked against `capacity`.
+     * If enough room exists for the request, returns success as true and tokens as how many tokens remain in the current fixed window.
+     *
+     * If a window does exist in the cache, we first check if the timestamp is greater than the fixedWindowStart + windowSize.
+     * If it isn't, we update currentToken with the incoming token until reach the capcity
+     *
+     * @param {string} uuid - unique identifer used to throttle requests
+     * @param {number} timestamp - time the request was recieved
+     * @param {number} [tokens=1]  - complexity of the query for throttling requests
+     * @return {*}  {Promise<RateLimiterResponse>}
+     * @memberof FixedWindow
+     */
+    async processRequest(
+        uuid: string,
+        timestamp: number,
+        tokens = 1
+    ): Promise<RateLimiterResponse> {
+        // set the expiry of key-value pairs in the cache to 24 hours
+        const keyExpiry = 86400000;
+
+        // attempt to get the value for the uuid from the redis cache
+        const windowJSON = await this.client.get(uuid);
+
+        if (windowJSON === null) {
+            const newUserWindow: RedisWindow = {
+                currentTokens: tokens > this.capacity ? 0 : tokens,
+                fixedWindowStart: timestamp,
+            };
+
+            if (tokens > this.capacity) {
+                await this.client.setex(uuid, keyExpiry, JSON.stringify(newUserWindow));
+                return { success: false, tokens: this.capacity };
+            }
+            await this.client.setex(uuid, keyExpiry, JSON.stringify(newUserWindow));
+            return { success: true, tokens: this.capacity - newUserWindow.currentTokens };
+        }
+        const window: RedisWindow = await JSON.parse(windowJSON);
+
+        // update the currentToken until reaches its capacity
+        window.currentTokens += tokens;
+        if (window.currentTokens > this.capacity) {
+            window.currentTokens = this.capacity;
```

Comment:
You'll need to check for a new time window before this part, because even if the users `currentTokens + tokens` is greater than `capacity`, if they are in a new window the request could be ok.

Reply:
You could move the `updateTimeWindow` function to before this part, to correctly update the window to what it should be base on the `timestamp`, then check if `tokens` or `currentTokens + tokens` is greater than capacity

Reply:
Updated in the latest version.